src/GitVersionTask/GitVersionTaskBase.cs

```diff
@@ -1,72 +1,141 @@
 
-    using Microsoft.Build.Framework;
-    using Microsoft.Build.Utilities;
-
-    public abstract class GitVersionTaskBase : Task
+    public static class GitVersionTaskBase
     {
-        protected GitVersionTaskBase()
+        internal static TOutput ExecuteGitVersionTask<TInput, TOutput>(
+            TInput input,
+            Func<TInput, TaskLogger, TOutput> execute
+            )
+            where TInput : AbstractInput
+            where TOutput : class, new()
         {
-            var fileSystem = new FileSystem();
-            ExecuteCore = new ExecuteCore(fileSystem);
-            GitVersion.Logger.SetLoggers(LogDebug, LogInfo, LogWarning, s => LogError(s));
-        }
+            if (!input.ValidateInput())
+            {
+                throw new Exception("Invalid input.");
+            }
 
-        public override bool Execute()
-        {
+            var logger = new TaskLogger();
+            Logger.SetLoggers(logger.LogInfo, logger.LogInfo, logger.LogWarning, s => logger.LogError(s));
+
+
+            TOutput output = null;
             try
             {
-                InnerExecute();
-                return true;
+                output = execute(input, logger);
             }
             catch (WarningException errorException)
             {
-                LogWarning(errorException.Message);
-                return true;
+                logger.LogWarning(errorException.Message);
+                output = new TOutput();
             }
             catch (Exception exception)
             {
-                LogError("Error occurred: " + exception);
-                return false;
+                logger.LogError("Error occurred: " + exception);
+                throw;
             }
+            finally
+            {
+                Logger.Reset();
+            }
+
+            return output;
         }
 
-        protected abstract void InnerExecute();
 
-        protected ExecuteCore ExecuteCore { get; }
+        public static ExecuteCore CreateExecuteCore()
+            => new ExecuteCore(new FileSystem());
+
+        private static string GetFileExtension(this String language)
+        {
+            switch (language)
+            {
+                case "C#":
+                    return "cs";
 
-        [Required]
-        public string SolutionDirectory { get; set; }
+                case "F#":
+                    return "fs";
 
-        public bool NoFetch { get; set; }
+                case "VB":
+                    return "vb";
 
-        public void LogDebug(string message)
-        {
-            BuildEngine.LogMessageEvent(new BuildMessageEventArgs(message, string.Empty, "GitVersionTask", MessageImportance.Low));
+                default:
+                    throw new Exception($"Unknown language detected: '{language}'");
+            }
         }
 
-        public void LogWarning(string message)
+        public static FileWriteInfo GetWorkingDirectoryAndFileNameAndExtension(
```

Comment:
I think this is a bit hard to read. Is it possible to make this into a class with delegates or something more descriptive, using less words to describe what it does?